### PR TITLE
add support for a hal remote library

### PR DESCRIPTION
**Pull request by coldav**
# Overview

This adds support for generic client and server hal support to allow
sending commands as binary requests over a `hal_transmitter`.
    
# Reason for change

Required to support customer needs of talking to a remote cpu.

# Description of change

Added encoders and decoders for converting hal commands to binary format
    
Add hal_server base class which uses the encoders and decoders and a
hal_transmitter to receive hal commands remotely and act on the given
hal device.
    
Add a hal_client and hal_device_client base class which can be used with
a transmitter to send to a remote server binary commands.
    
Added a hal socket cpu client.
Added hal cpu socket based server program.

# Anything else we should know?

This currently only supports single device hals

### CODE REVIEW

File: examples/hals/hal_cpu_client/source/hal_cpu_executable.cpp
Comment (hvdijk) [inline examples/hals/hal_cpu_client/source/hal_cpu_executable.cpp:35]: This should probably go to `stderr`.

File: examples/hals/hal_cpu_client/source/hal_cpu_executable.cpp
Comment (hvdijk) [inline examples/hals/hal_cpu_client/source/hal_cpu_executable.cpp:51]: Are we sure that after `status_transmitter_failed`, the code is in a state where the server can be meaningfully restarted? I would personally prefer to have a service manager restart the whole process automatically (several service managers exist) but I'm okay with this approach if we know the process is in a good enough state.

File: examples/hals/hal_cpu_client/source/hal_main.cpp
Comment (hvdijk) [inline examples/hals/hal_cpu_client/source/hal_main.cpp:41]: This lock doesn't do anything, does it?

File: hal/hal_remote/CMakeLists.txt
Comment (hvdijk) [inline hal/hal_remote/CMakeLists.txt:28]: On non-Linux systems, what is the result? Do we get a `hal_remote` library with no object files?

File: hal/hal_remote/include/hal_remote/hal_binary_decoder.h
Comment (hvdijk) [inline hal/hal_remote/include/hal_remote/hal_binary_decoder.h:95]: This is unreachable.

File: hal/hal_remote/include/hal_remote/hal_binary_decoder.h
Comment (hvdijk) [inline hal/hal_remote/include/hal_remote/hal_binary_decoder.h:98]: Since this is data from an untrusted client, we probably can't do `assert(0)`, we probably have to take that out and always do the `return`?

File: hal/hal_remote/include/hal_remote/hal_client.h
Comment (hvdijk) [inline hal/hal_remote/include/hal_remote/hal_client.h:65]: This is wrong if server and client disagree on endianness. We only have little endian targets at the moment, can we add an assert for that?

File: hal/hal_remote/include/hal_remote/hal_binary_decoder.h
Comment (hvdijk) [inline hal/hal_remote/include/hal_remote/hal_binary_decoder.h:109]: Can we use `cargo::result` here to make it clearer from the API what the meaning of the result is?

Comment (hvdijk) [reply hal/hal_remote/include/hal_remote/hal_binary_decoder.h:109]: Also, I cannot really follow the logic of when `uint8_t *` is used vs. `void *`, I would have expected `void *` here but it's not a problem if there is a reason for it.

File: examples/hals/hal_cpu_client/source/hal_cpu_executable.cpp
Comment (coldav) [reply examples/hals/hal_cpu_client/source/hal_cpu_executable.cpp:51]: it's a bit convoluted but I check the connection_closed above so if it fails at this point it's because the connection is closed. However I think we agreed before that this could lead to all sorts of issues with allocs being left over etc, so I'll take out the loop and advise the customer to use a bash while loop as an option.

File: hal/hal_remote/CMakeLists.txt
Comment (coldav) [reply hal/hal_remote/CMakeLists.txt:28]: when I had it initially as all headers, it didn't seem to complain, but it does have some .cpp files above anyway (L24).

File: hal/hal_remote/include/hal_remote/hal_binary_decoder.h
Comment (coldav) [reply hal/hal_remote/include/hal_remote/hal_binary_decoder.h:109]: I don't want to drag in cargo, we've avoided it in hal so far in keeping it mostly standalone (and working with clik).
I think you are right about void * though, although we do tend to need to then cast it elsewhere to index into it.

File: hal/hal_remote/include/hal_remote/hal_client.h
Comment (coldav) [reply hal/hal_remote/include/hal_remote/hal_client.h:65]: I agree it depends on matching endianness. I'll have to have a think about best to check the endianness with an assert (in theory it's in the info).

File: examples/hals/hal_cpu_client/source/hal_main.cpp
Comment (coldav) [reply examples/hals/hal_cpu_client/source/hal_main.cpp:41]: nothing useful no. I will remove.

File: examples/hal_cpu_remote_server/CMakeLists.txt
Comment (hvdijk) [inline examples/hal_cpu_remote_server/CMakeLists.txt:20]: Can we move `examples/clik/external/hal_cpu` to `examples/external/hal_cpu` since it is no longer limited to `clik`?

File: examples/hal_cpu_remote_server/README.md
Comment (hvdijk) [inline examples/hal_cpu_remote_server/README.md:50]: I could be misremembering, but didn't you want to change this process before merging?

File: examples/hal_cpu_remote_server/README.md
Comment (hvdijk) [inline examples/hal_cpu_remote_server/README.md:103]: We talked about mentioning the option of sandboxing in some way too here.

File: examples/hal_cpu_remote_server/hal_cpu_executable.cpp
Comment (hvdijk) [inline examples/hal_cpu_remote_server/hal_cpu_executable.cpp:56]: `?` is not a good option for `getopt` because the return value of `'?'` for an option `-?` is indistinguishable from the return value of `'?'` for an unrecognised option, and the option `-?` is not something we should try to support since `?` is a shell metacharacter and so unless quoted, this looks for files in the current directory whose names are `-` followed by a single character.

File: examples/hal_cpu_remote_server/README.md
Comment (coldav) [reply examples/hal_cpu_remote_server/README.md:103]: do you have any suggestions e.g. You may also consider running in a docker to sandbox it?

File: examples/hal_cpu_remote_server/CMakeLists.txt
Comment (coldav) [reply examples/hal_cpu_remote_server/CMakeLists.txt:20]: That's a bigger change since the original idea of clik was that it was fairly standalone. I think it's a good idea, but I'd prefer to to do it in a separate PR.

File: examples/hal_cpu_remote_server/README.md
Comment (coldav) [reply examples/hal_cpu_remote_server/README.md:50]: I think it's a big enough change, we should consider doing it separately and not hold up this PR.

File: hal/include/hal_types.h
Comment (hvdijk) [inline hal/include/hal_types.h:172]: We should be detecting this rather than assuming, I think.

File: examples/hal_cpu_remote_server/README.md
Comment (hvdijk) [reply examples/hal_cpu_remote_server/README.md:103]: Could do that, or even just running under a user account that does not have access to anything.

Comment (coldav) [reply examples/hal_cpu_remote_server/README.md:103]: I already say that above, I've also added dockers as a mention.

File: hal/include/hal_types.h
Comment (coldav) [reply hal/include/hal_types.h:172]: This is a default in the base Hal, it's up to the writer of each HAL to dictate this value (also because clc uses this an offline way, we can't easily detect other than the host) - I have added some additional checks on the host side though to see that it has the same reported endianness.

Comment (hvdijk) [reply hal/include/hal_types.h:172]: That makes sense, being unable to detect it. But it still shouldn't be necessary to set it here. I see it's already getting set as it should be in `cpu_hal::setup_cpu_hal_device_info`.

Comment (coldav) [reply hal/include/hal_types.h:172]: This is more about avoiding uninitialized variables than it being the defacto value.

Comment (hvdijk) [reply hal/include/hal_types.h:172]: But it gets initialized to `true` rather than `{}`, and other fields are not initialized similarly.

Comment (coldav) [reply hal/include/hal_types.h:172]: fair enough, I'll remove it, I've just seen issues with not setting things to default values here, especially when we add fields, but I'll think on it and remove it from this change.

Comment (coldav) [reply hal/include/hal_types.h]: removed.

File: examples/hal_cpu_remote_server/hal_cpu_executable.cpp
Comment (hvdijk) [reply examples/hal_cpu_remote_server/hal_cpu_executable.cpp:63]: This does print to `stderr` on bad options now, but still exits with exit code 0.

File: examples/hal_cpu_remote_server/hal_cpu_executable.cpp
Comment (hvdijk) [inline examples/hal_cpu_remote_server/hal_cpu_executable.cpp:64]: Nit: code after `exit(0)` isn't going to execute :)

**Review by hvdijk — APPROVED** [review]

**Pull request merged**
